### PR TITLE
Fail CommandSequence if RESTful call fails

### PR DESCRIPTION
**Pull request by vladak**
This changes the processing of RESTful calls in `CommandSequence`. If the call fails, the "command" is considered as failed.

### CODE REVIEW

File: opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py [hunk 1]
```diff
@@ -60,7 +64,7 @@ def __str__(self):
 
     def get_cmd_output(self, cmd, indent=""):
         str = ""
-        if self.outputs[cmd]:
+        if self.outputs.get(cmd):
             for line in self.outputs[cmd]:
                 str += '{}{}'.format(indent, line)
```
Comment (ericsaintetienne) [inline opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py:69]:
```suggestion
        for line in self.outputs.get(cmd, []):
            str += '{}{}'.format(indent, line)
```

File: opengrok-tools/src/main/python/opengrok_tools/utils/webutil.py [hunk 2]
```diff
@@ -27,45 +27,23 @@
 
 
 def get(logger, uri, params=None, headers=None):
-    try:
-        proxies = get_proxies(uri)
-        return requests.get(uri, params=params, proxies=proxies)
-    except Exception:
-        logger.debug(traceback.format_exc())
-        return None
+    proxies = get_proxies(uri)
```
Comment (ericsaintetienne) [inline opengrok-tools/src/main/python/opengrok_tools/utils/webutil.py:30]: A lot of repetitions here, wouldn't it make sense to factor the common code into a single function (named, say `request`) that all verbs would call?

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/utils/webutil.py]: refactored. I could possibly go as far as making `do_api_call()` to use `requests.__dict__` to lookup the function names.

**Comment by vladak** [comment]
Thanks @ericsaintetienne for taking a look !

**Comment by vladak** [comment]
Tested with invalid JSON:
```json
{"messageFubar": "warning", "duration": "PT1H", "tags": ["foo"], "text": "resync + reindex in progress"}
```
and `opengrok-sync` config:
```yml
commands:
- command: [echo, 'foo']
- command:
  - http://localhost:8080/source/api/v1/messages
  - POST
  - messageFubar: warning
    duration: PT1H
    tags: ['%PROJECT%']
    text: resync + reindex in progress
```
When run, the command reports the error and exits with non zero exit code:
```
./env/bin/opengrok-sync -c ../sync-bad_restful.yml 
Assuming directory: /var/opengrok/src
RESTful command {'command': ['http://localhost:8080/source/api/v1/messages', 'POST', {'messageFubar': 'warning', 'duration': 'PT1H', 'tags': ['%PROJECT%'], 'text': 'resync + reindex in progress'}]} failed: 400 Client Error:  for url: http://localhost:8080/source/api/v1/messages
...
processing of project 'universal-ctags' failed
  failed commands:
    '{'command': ['http://localhost:8080/source/api/v1/messages', 'POST', {'messageFubar': 'warning', 'duration': 'PT1H', 'tags': ['%PROJECT%'], 'text': 'resync + reindex in progress'}]}': 1
$ echo $?
1
```

File: opengrok-tools/src/main/python/opengrok_tools/utils/restful.py [hunk 3]
```diff
@@ -18,31 +18,27 @@
...
-        'PUT': put,
-        'POST': post,
-        'DELETE': delete
-    }
-    handler = verbs.get(verb)
-    if handler is not None:
-        logger = logging.getLogger(__name__)
-        return handler(logger, uri, headers=headers, data=data)
-    raise Exception('Unknown HTTP verb in command {}'.format(command))
+def do_api_call(uri, verb, headers=None, data=None):
```
Comment (tulinkry) [inline opengrok-tools/src/main/python/opengrok_tools/utils/restful.py:36]: I believe you can pass the handler directly as `do_api_call('http:/...', requests.put, ... )` which could simplify the code around it

Comment (tulinkry) [reply opengrok-tools/src/main/python/opengrok_tools/utils/restful.py:36]: Also I would switch the order to verb -> uri as that is the usual order

File: opengrok-tools/src/main/python/opengrok_tools/sync.py [hunk 4]
```diff
@@ -109,7 +111,11 @@ def do_sync(args, commands, config, directory, dirs_to_process, ignore_errors,
                 cmds = CommandSequence(cmds_base)
                 cmds.fill(cmds_base.retcodes, cmds_base.outputs,
                           cmds_base.failed)
-                cmds.check(ignore_errors)
+                r = cmds.check(ignore_errors)
+                if r != 0:
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/sync.py]: yep

File: opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py [hunk 5]
```diff
@@ -156,7 +165,11 @@ def run_cleanup(self):
 
         for cleanup_cmd in self.cleanup:
             if is_web_uri(cleanup_cmd.get(COMMAND_PROPERTY)[0]):
-                call_rest_api(cleanup_cmd, PROJECT_SUBST, self.name)
+                try:
+                    call_rest_api(cleanup_cmd, PROJECT_SUBST, self.name)
+                except HTTPError as e:
+                    self.logger.error("RESTful command {} failed: {}".
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py:171]: should be fixed.

File: opengrok-tools/src/main/python/opengrok_tools/utils/restful.py (hunk 3 above)
Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/utils/restful.py:36]: Passing the handler directly would be arguably more error prone + would need a dependency in callers + is longer to type.

Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/utils/restful.py:36]: changed the signature to use verb first.

File: opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py [hunk 6]
```diff
@@ -106,12 +109,18 @@ def run(self):
...
         for command in self.commands:
             if is_web_uri(command.get(COMMAND_PROPERTY)[0]):
-                call_rest_api(command, PROJECT_SUBST, self.name)
+                try:
+                    call_rest_api(command, PROJECT_SUBST, self.name)
+                except HTTPError as e:
+                    self.logger.error("RESTful command {} failed: {}".
+                                      format(command, e))
+                    self.failed = True
+                    self.retcodes[str(command)] = FAILURE_EXITVAL
```
[the remark this answers is not in the document]
Comment (vladak) [reply opengrok-tools/src/main/python/opengrok_tools/utils/commandsequence.py:123]: good catch. The RESTful calls should be equally handled.

**Review by tulinkry — APPROVED** [review]

**Comment by vladak** [comment]
Thanks @tulinkry !

**Pull request merged**